Index: funkload/FunkLoadTestCase.py
===================================================================
--- funkload/FunkLoadTestCase.py	(revision 51244)
+++ funkload/FunkLoadTestCase.py	(working copy)
@@ -201,7 +201,8 @@
             self.total_links += 1
         if rtype in ('post', 'get', 'redirect'):
             # this is a valid referer for the next request
-            self.setHeader('Referer', url)
+            #self.setHeader('Referer', url)
+            pass
         self._browser.history.append((rtype, url))
         self.logd(' Done in %.3fs' % t_delta)
         self._log_response(response, rtype, description, t_start, t_stop)
